crates/bevy_window/src/window.rs

```diff
@@ -23,20 +23,40 @@ pub struct WindowId(Uuid);
 #[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
 #[doc(alias = "vsync")]
 pub enum PresentMode {
+    /// Chooses FifoRelaxed -> Fifo based on availability.
+    ///
+    /// Because of the fallback behavior, it is supported everywhere.
+    AutoVsync = 0,
+    /// Chooses Immediate -> Mailbox -> Fifo (on web) based on availability.
+    ///
+    /// Because of the fallback behavior, it is supported everywhere.
+    AutoNoVsync = 1,
     /// The presentation engine does **not** wait for a vertical blanking period and
     /// the request is presented immediately. This is a low-latency presentation mode,
     /// but visible tearing may be observed. Will fallback to `Fifo` if unavailable on the
     /// selected platform and backend. Not optimal for mobile.
-    Immediate = 0,
+    ///
+    /// Selecting this variant will panic if not supported, it is preferred to use
+    /// [`PresentMode::AutoNoVsync`].
+    #[deprecated = "May panic if unsupported. Use PresentMode::AutoNoVsync"]
+    Immediate = 2,
     /// The presentation engine waits for the next vertical blanking period to update
     /// the current image, but frames may be submitted without delay. This is a low-latency
     /// presentation mode and visible tearing will **not** be observed. Will fallback to `Fifo`
     /// if unavailable on the selected platform and backend. Not optimal for mobile.
-    Mailbox = 1,
+    ///
+    /// Selecting this variant will panic if not supported, it is preferred to use
+    /// [`PresentMode::AutoNoVsync`].
+    #[deprecated = "May panic if unsupported. Use PresentMode::AutoNoVsync"]
+    Mailbox = 3,
```

Comment:
Why is this being deprecated? There's no other way to select vsync-without-tearing, aka "fast vsync"? This is often the mode you want on desktop to allow for low input latency (and external techniques can be used for frame throttling). 

I can understand making one of the Auto variants the default, but removing the ability for fine grain control here seems like a step backwards.

Reply:
Wgpu doesn't fallback anymore, you must request a supported mode (with Fifo _always_ supported)

Otherwise I agree, these shouldn't be deprecated.

Reply:
Yeah, I think deprecating these is the wrong approach. And I think `AutoVsync` preferring `FifoRelaxed`, which can tear when fps < vsync, is undesirable.

One idea is to allow the user to specify a callback in `WindowSettings` that takes a `&[PresentMode]` and returns a `PresentMode`. That way users can implement a callback that attempts to choose a supported present mode and they can use whatever logic they like to choose what to use. Maybe the AdapterInfo should also be an argument there. Another is to try to use the chosen `PresentMode` and if they chose `Immediate` or `Mailbox` and it isn't supported, fall back to `Fifo` and log a warning.

I tried to look over the situation with support and the parameters one tries to prioritise when choosing the `PresentMode` and spammed out my thoughts a bit on Discord. My understanding basically lands like this:

![wgpu PresentMode support table showing the different present modes versus the API/GPU vendor support as documented in the wgpu PresentMode documentation](https://user-images.githubusercontent.com/302146/177987655-f3ff573d-7a58-4262-9b19-5643d5f7ede9.jpg)

The parameters of interest are: tearing, input/simulation to photon latency, energy/power consumption.

My summary of the different `PresentModes` is:
- Immediate
    - lowest latency (as the presented frame is immediately swapped)
    - tears
- FifoRelaxed
    - lower latency than Fifo at fps lower than vsync
    - tears at fps lower than vsync
- Mailbox
    - potentially lower latency than fifo at fps higher than vsync (potentially if no frame pacing is used)
    - does more rendering work than is displayed at fps higher than vsync
    - no tearing
- Fifo
    - potentially higher latency than other options when fps is higher than vsync (potentially if no frame pacing is used)
    - only does as much rendering work as is actually displayed
    - frame pacing can reduce latency
    - no tearing

And then given the following:
- frame time (as in CPU and GPU processing time) can be measured separately from being blocked waiting for anything, so that actual time spent doing work is measurable
  - we use `Immediate` in stress tests to enable the frame rate diagnostics to show 'actual' frame rate though this is not the most reliable measure anyway. I always prefer to run a profile, but I do use `Immediate` mode. `Immediate` mode would not matter if we could measure the time spent doing work and not the time spent waiting, which we can because we can measure the time spent waiting for `get_current_texture()`
- simulation and input event handling is decoupled from the render loop or otherwise delayed so that input events are processed as late before rendering and presentation as possible using frame pacing
- screen refresh is trending beyond 60Hz to >= 120Hz

...I'm struggling to see why anyone should choose anything other than `Fifo`. Bear with me...

- `Mailbox` has no benefit if frame pacing is used, unless I'm missing something. It just does more work than necessary (as rendered frames would be discarded if fps > vsync.)
- I don't know why anyone would want a mode that can tear. Maybe some idea of lower latency because the image is swapped out immediately, but it only means that you would see part of an image sooner as it was replaced while the framebuffer was being scanned out to the screen, but that could just as easily cause more mental confusion than any latency benefit as you're looking at 2 or more time snapshots on-screen.
- Frame pacing and using presentation times for animation/simulation should give a much smoother and more predictable motion and low best case input to photon latency. Being able to predict motion is important for gameplay.
- I suppose with a 60Hz display there's a fair amount of time between processing input for frames but if that's important, it's usually for competitive purposes and then as mentioned, processing input and conducting simulation could be done at a higher rate if you cannot afford/obtain a 120+Hz display.

So `Immediate` and `Mailbox` are only beneficial for development if you can't measure frame times accurately, and only beneficial for gameplay if frame pacing does not work well and/or if input event processing and simulation are limited by vsync. These feel like problems we can do a good job at 'solving', providing good defaults (frame pacing) and options where relevant (decoupled input event handling / simulation for the purposes of higher rate of interactivity than display). Maybe that's optimistic. I do feel like frame pacing is going to be a hard problem to solve unless people (bevy developers and users of bevy developing their own games) adhere to frame time budgets as otherwise if frame time varies significantly then predictability goes out the window.

And to be clear, I think we should definitely give users control and the ability to choose what priority they want. I don't think just `AutoVsync` and `AutoNoVsync` will cut it.

Reply:
I put the deprecated after having something that was working in native crash in wasm. My idea was indeed not to deprecate those modes, but to have an indication that they are not the recommended one anymore. Is there another annotation that would display a warning when they are used?

Reply:
> I put the deprecated after having something that was working in native crash in wasm. My idea was indeed not to deprecate those modes, but to have an indication that they are not the recommended one anymore. Is there another annotation that would display a warning when they are used?

That's not the purpose of `#[deprecated]` though. Having something marked deprecated indicates that it's obsolete and will be removed in the future. In any case, all variants are valid. I think this should be addressed with documentation and the default should be `Fifo`. I would be surprised to have a default setting that allows tearing.

Reply:
Yup I'll remove the deprecated, I'm waiting on a wgpu update to finish this PR